### ISSUE TITLE
Why generate/store tree nodes in bottom-up order?

### ISSUE THREAD

**Issue by bjornharrtell** [mourner/flatbush#25]
As noted:
https://github.com/mourner/flatbush/blob/22656c4cee2e489be7e0f203e2d824cd36898048/index.js#L127

Why I ask is, if you utilize the flatbush structure for very large indexes and you are interested in doing a tree search remotely via network or other high-latency it will require fetching the nodes backwards from the end which is an I/O pattern not usually optimized for using prefetch/cache techniques and even more problematic for sources that cannot seek backwards efficiently.

For reference these thoughts was triggered by observations on the flatbush structure as implemented in FlatGeobuf, tracked by https://github.com/bjornharrtell/flatgeobuf/issues/39.

**Comment by bjornharrtell** [comment]
Might have my answer at https://www.ibm.com/support/knowledgecenter/SSGU8G_12.1.0/com.ibm.spatial.doc/ids_spat_075.htm.

**Comment by mourner** [comment]
This is a good question! The link above is about a different thing — it talks about bottom-up vs top-down algorithm for building the tree index (regardless of how nodes are laid out), while the original post is more about the order in which nodes are stored in the array. I did it this way because it was straightforward in implementation and I didn't consider other contexts than JavaScript (where order of access doesn't make much difference in performance).

However it's totally reasonable to reverse the order so that the root is at the start and nodes follow in the order from top nodes to leaves at the end. We would just need to change how generated nodes are saved in the pre-allocated array. I wonder how much this would impact search (which would need to adapt accordingly to go from left to right). If you're up to a PR on this, let's see how that works out!

**Comment by bjornharrtell** [comment]
Thanks, was thinking along those lines but unsure if there was some detail in the storage order and search algorithm that made it unfeasible.

I actually have a streaming variant of the search algorithm in TypeScript at https://github.com/bjornharrtell/flatgeobuf/blob/3221fa5c1ac0d22616975a667448a71f10e4f8ad/src/ts/packedrtree.ts#L48-L99 so this can be relevant even for browser/JavaScript use. Looking at that again I realize there is also the issue that accesses for indices and bboxes are interleaved.

**Comment by bjornharrtell** [comment]
Looks like I managed to do it in my C++ fork. As far as I can tell I get completely sequential traversal and in the end the changes was not that large. However, to get sequential I have included the indices in the node items and that might be too invasive of a change for flatbush (will perhaps not be possible to use typed arrays as efficiently) and might also reduce performance for the case it was made/optimized for i.e indexes that fit in RAM and typed arrays for maximum performance.
